src/System.Linq/src/System/Linq/Skip.cs

```diff
@@ -15,21 +15,23 @@ public static IEnumerable<TSource> Skip<TSource>(this IEnumerable<TSource> sourc
                 throw Error.ArgumentNull(nameof(source));
             }
 
-            if (count < 0)
+            IList<TSource> sourceList = source as IList<TSource>;
+            if (sourceList != null)
             {
-                count = 0;
+                return new ListPartition<TSource>(sourceList, Math.Max(0, count), int.MaxValue);
             }
 
-            IPartition<TSource> partition = source as IPartition<TSource>;
-            if (partition != null)
+            if (count <= 0)
             {
-                return partition.Skip(count);
+                // If source is an Iterator or implements IPartition then it's internal to this assembly and came from here, so it
+                // won't be the same instance as something that wasn't from a Linq method.
+                return source as Iterator<TSource> ?? source as IPartition<TSource> ?? Wrap(source);
```

Comment:
I'm not convinced this case is impactful enough to warrant introducing yet another custom partition implementation.